flow/connectors/clickhouse/normalize.go

```diff
@@ -278,170 +280,202 @@ func (c *ClickHouseConnector) NormalizeRecords(
 		return nil, err
 	}
 
-	rawTbl := c.getRawTableName(req.FlowJobName)
+	parallelNormalize, err := peerdbenv.PeerDBClickHouseParallelNormalize(ctx, req.Env)
+	if err != nil {
+		return nil, err
+	}
 
-	// model the raw table data as inserts.
-	for _, tbl := range destinationTableNames {
-		// SELECT projection FROM raw_table WHERE _peerdb_batch_id > normalize_batch_id AND _peerdb_batch_id <= sync_batch_id
-		selectQuery := strings.Builder{}
-		selectQuery.WriteString("SELECT ")
+	var destinationTableNameChunks iter.Seq[[]string]
+	if parallelNormalize > 1 {
+		var chunkSize int
+		if parallelNormalize >= len(destinationTableNames) {
+			chunkSize = 1
+		} else {
+			chunkSize = shared.DivCeil(len(destinationTableNames), parallelNormalize)
+		}
+		destinationTableNameChunks = slices.Chunk(destinationTableNames, chunkSize)
+	} else {
+		destinationTableNameChunks = func(yield func([]string) bool) {
+			yield(destinationTableNames)
+		}
+	}
 
-		colSelector := strings.Builder{}
-		colSelector.WriteRune('(')
+	enablePrimaryUpdate, err := peerdbenv.PeerDBEnableClickHousePrimaryUpdate(ctx, req.Env)
+	if err != nil {
+		return nil, err
+	}
 
-		schema := req.TableNameSchemaMapping[tbl]
+	rawTbl := c.getRawTableName(req.FlowJobName)
 
-		var tableMapping *protos.TableMapping
-		for _, tm := range req.TableMappings {
-			if tm.DestinationTableIdentifier == tbl {
-				tableMapping = tm
-				break
+	group, errCtx := errgroup.WithContext(ctx)
+	for tableNames := range destinationTableNameChunks {
```

Comment:
can we log the table names that we are processing here as well please?

Reply:
have revised system to run query at a time, query gets logged